### ISSUE TITLE
DR-DOS 7.02 doesn't boot (7.03 works, 7.02 doesn't)

### ISSUE THREAD

**Issue by jschwartzenberg** [stsp/dosemu2#151]
With current GIT, I can't get DR-DOS 7.03 to boot on an AMD64 system (Ubuntu 14.04). I'm installing it with my getdrdos script from here:
https://github.com/jschwartzenberg/dos-download-scripts/

When starting it with KVM, I get:
```
ERROR: unexpected CPU exception 0x06 err=0x00000000 cr2=40fbe000 while in vm86 (DOS)
Program=do_vm86.c, Line=158
EIP: 0070:000074f8 ESP: 0000:00007bf4  VFLAGS(b): 01011 00000010 01000110
EAX: 00000000 EBX: 00000256 ECX: 0000f800 EDX: 00000080 VFLAGS(h): 000b0246
ESI: 00000000 EDI: 00000000 EBP: 00000000 DS: 0000 ES: 0000 FS: 0000 GS: 0000
FLAGS: PF ZF IF RF VM VIF  IOPL: 0
STACK: 00 00 00 00 00 00 00 00 70 00 -> 78 00 00 00 c7 6f 00 f8 00 00 
OPS  : 00 00 00 00 70 00 78 00 00 00 -> c7 6f 00 f8 00 00 80 00 eb 3e 
        c76f00f800          0070:74f8 mov  word [bx],00F8
```

When starting it with emulated, I get:
```
dosemu.bin: do_vm86.c:627: callback_return: Assertion `callback_level > 0' failed.
```

**Comment by stsp** [comment]
And when did it work?

**Comment by stsp** [comment]
Or with what setting?

**Comment by jschwartzenberg** [comment]
No idea if it ever worked, I'll try it with the 1.5.0 fork point version and on 32-bit and see what happens

**Comment by andrewbird** [comment]
According to my issue #125 it was booting back in December.

**Comment by stsp** [comment]
Can someone please bisect? :)

**Comment by jschwartzenberg** [comment]
Could it be you had another version of DR-DOS? I went back to 3e0af21559c18acd880ecd6b6dc186a6763da533 but even with that version it doesn't boot (32-bit).

**Comment by andrewbird** [comment]
I have the sha1sums of the three boot files, but it will take a few minutes to get them.

**Comment by jschwartzenberg** [comment]
Here are mine:
```
5757dc7604e12fb3bbfe96e53a37028060bcedf3  COMMAND.COM
795faf2e5d936cabe27127c1d8379adb15c791d1  IBMBIO.COM
39ed49c4b6c3641061087b0b799c42fbc41b129c  IBMDOS.COM
```

**Comment by andrewbird** [comment]
Here are the sha1sums from my test suite:
```Python
        cls.version = "Caldera DR-DOS 7.03"                                     
        cls.tarfile = "DR-DOS-7.03.tar"                                         
        cls.files = [                                                           
            ("ibmbio.com", "a2b588314f331a93ce2ddbe63e4e9ed19b5dc864"),         
            ("ibmdos.com", "e89e9c1f956c8249864fdb42037f29061729b709"),         
            ("command.com", "90eca5bdcc851fd0b977bf6dbbab040b4797cc2c"),        
        ]
```
~~~
ajb@polly:/clients/common/dosemu2.git$ nosetests -v test/test_bootos.py:DRDOS703TestCase
Test DR-DOS-7.03 FAT12 ... ok
Test DR-DOS-7.03 FAT16 ... ok
Test DR-DOS-7.03 FAT16B ... ok
Test DR-DOS-7.03 VBFLPY ... ok

----------------------------------------------------------------------
Ran 4 tests in 12.714s

OK
~~~


Since I need to sort out the branch / rebase, I can't test at the moment against current devel, but that test run was against (plus some local patches that I doubt affect anything)
~~~
commit 2de93722c51b82300de5e2c17fcc6eac9c23f13b
Author: Stas Sergeev <stsp@users.sourceforge.net>
Date:   Fri Feb 26 16:59:30 2016 +0300
~~~

Environment is 32bit, cpuemu=default (which I think is off)

**Comment by andrewbird** [comment]
Mmm, it seems there are at least two versions of DR-DOS 7.03!

**Comment by jschwartzenberg** [comment]
It turned out that I actually had version 7.02. Downloading 7.03 properly still gives me different sha1sums though (and it doesn't work):
```
5757dc7604e12fb3bbfe96e53a37028060bcedf3  COMMAND.COM
795faf2e5d936cabe27127c1d8379adb15c791d1  IBMBIO.COM
39ed49c4b6c3641061087b0b799c42fbc41b129c  IBMDOS.COM
```
I'm extracting the files from this image:
http://archive.org/download/IBM_PC_Compatibles_TOSEC_2012_04_23/IBM_PC_Compatibles_TOSEC_2012_04_23.zip/IBM%20PC%20Compatibles%20%5BTOSEC%5D%2FIBM%20PC%20Compatibles%20-%20Operating%20Systems%20-%20%5BIMA%5D%20%28TOSEC-v2011-01-06_CM%29%2FCaldera%20DR-DOS%20v7.03%20%281998%29%28Caldera%20Systems%29%28Disk%201%20of%205%29.zip

**Comment by andrewbird** [comment]
Mine came from winworldpc.com and was called 'Digital Research DR-DOS 7.03 (3.5)'  at [link](https://winworldpc.com/download/410219DF-18DA-11E4-99E5-7054D21A8599)

**Comment by andrewbird** [comment]
7.02 is not working for me either and results in:
~~~
SIGILL while in vm86(): 0000:7c66                                               
ERROR: unexpected CPU exception 0x06 err=0x00000000 cr2=00000000 while in vm86 (DOS)
~~~

**Comment by andrewbird** [comment]
Mmm, the two sets of sha1sums you posted are identical, are they for 7.02 or 7.03?

**Comment by jschwartzenberg** [comment]
Whoops, you're right. My bad. Those were for 7.02. I just got 7.03 and it works. It seems that it's just 7.02 which doesn't work then.

**Comment by andrewbird** [comment]
Actually to answer my own question, the 7.03 zip image you linked is identical to mine
~~~
root@polly:~# sha1sum /a/IBM* /a/COMMAND.COM 
a2b588314f331a93ce2ddbe63e4e9ed19b5dc864  /a/IBMBIO.COM
e89e9c1f956c8249864fdb42037f29061729b709  /a/IBMDOS.COM
90eca5bdcc851fd0b977bf6dbbab040b4797cc2c  /a/COMMAND.COM
~~~

Note: I just saw your post - glad we agree!

**Comment by andrewbird** [comment]
Actually I'm short on time at the moment, so if anyone wants to fix 7.02 booting problem, please jump in!
I'd like to upload the test suite at some point, but I was waiting until I got
- 1/ a pure python solution (it currently uses 'timeout' system command to recover from hanging tests)
- 2/ more of the DR-DOS family booting properly (nobody likes tests that fail!)

The test suite currently doesn't have any binaries included as I doubt we'd be able to include those in the repo, so it tests for a tarfile with a standardised name, and either runs the test or skips it accordingly.

**Comment by stsp** [comment]
@jschwartzenberg Could you please test PR #158?

**Comment by jschwartzenberg** [comment]
Yep, that fixes DR-DOS 7.02!
